packages/engine/lib/error/src/macros.rs

```diff
@@ -0,0 +1,566 @@
+#[doc(hidden)]
+pub mod __private {
+    use core::fmt;
+
+    use crate::Report;
+
+    pub mod kinds {
+        use core::{fmt, marker::PhantomData};
+
+        use crate::Report;
+
+        pub trait AdhocKind: Sized {
+            fn __kind(&self) -> Adhoc {
+                Adhoc
+            }
+        }
+        impl<T> AdhocKind for &T where T: ?Sized + fmt::Display + fmt::Debug + Send + Sync + 'static {}
+
+        pub struct Adhoc;
+        impl Adhoc {
+            #[allow(clippy::unused_self)]
+            pub fn report<C>(self, context: C) -> Report
+            where
+                C: fmt::Display + fmt::Debug + Send + Sync + 'static,
+            {
+                Report::new(context)
+            }
+        }
+
+        pub trait TraitKind<S>: Sized {
+            fn __kind(&self) -> Trait<S> {
+                Trait(PhantomData)
+            }
+        }
+        impl<E, S> TraitKind<S> for E where E: Into<Report<S>> {}
+
+        pub struct Trait<S>(PhantomData<S>);
+        impl<S> Trait<S> {
+            #[allow(clippy::unused_self)]
+            pub fn report<E: Into<Report<S>>>(self, error: E) -> Report<S> {
+                error.into()
+            }
+        }
+    }
```

Comment:
Is the use of `S` here just because of Sized? Considering it's in `Report` I'd expect it to be `C`